src/AmplNLWriter.jl

```diff
@@ -1,934 +1,266 @@
-            filename,
-            "",
-            "",
-            NaN,
-            zeros(0),
-            :NotSolved,
-            -1,
-            -1,
-            "?",
-            "",
-            NaN,
-        )
-    end
-end
-mutable struct AmplNLLinearQuadraticModel <: AbstractLinearQuadraticModel
-    inner::AmplNLMathProgModel
-end
-mutable struct AmplNLNonlinearModel <: AbstractNonlinearModel
-    inner::AmplNLMathProgModel
-end
-
-include("nl_write.jl")
-
-function SolverInterface.NonlinearModel(s::AmplNLSolver)
-    return AmplNLNonlinearModel(
-        AmplNLMathProgModel(s.solver_command, s.options, s.filename),
+Create a new Optimizer object.
+
+`solver_command` should be one of two things:
+
+* A `String` of the full path of an AMPL-compatible executable
+* A function that takes takes a function as input, initializes any environment
+  as needed, calls the input function with a path to the initialized executable,
+  and then destructs the environment.
+
+# Examples
+
+A string to an executable:
+```julia
+Optimizer("/path/to/ipopt.exe", ["print_level=0"])
+```
+
+A function or string provided by a package:
+```julia
+Optimizer(Ipopt.amplexe, ["print_level=0"])
+# or
+Optimizer(Ipopt_jll.amplexe, ["print_level=0"])
+```
+
+A custom function
+```julia
+function solver_command(f::Function)
+    # Create environment ...
+    ret = f("/path/to/ipopt")
+    # Destruct environment ...
+    return ret
+end
+Optimizer(solver_command)
+```
+"""
+function Optimizer(
+    solver_command::Union{String,Function} = "",
+    options::Vector{String} = String[],
+)
+    model = Optimizer{Float64}()
+    model.ext[:VariablePrimalStart] = Dict{MOI.VariableIndex,Float64}()
+    model.ext[:NLPBlock] =
+        MOI.NLPBlockData(MOI.NLPBoundsPair[], _LinearNLPEvaluator(), false)
+    model.ext[:AMPLSolver] = _solver_command(solver_command)
+    model.ext[:Options] = options
+    model.ext[:Results] = _NLResults(
+        "Optimize not called.",
+        MOI.OPTIMIZE_NOT_CALLED,
+        MOI.NO_SOLUTION,
+        NaN,
+        Dict{MOI.VariableIndex,Float64}(),
     )
-end
-function SolverInterface.LinearQuadraticModel(s::AmplNLSolver)
-    return AmplNLLinearQuadraticModel(
-        AmplNLMathProgModel(s.solver_command, s.options, s.filename),
+    return model
+end
+
+function MOI.empty!(model::Optimizer{Float64})
+    model.name = ""
+    model.senseset = false
+    model.sense = MOI.FEASIBILITY_SENSE
+    model.objectiveset = false
+    model.objective =
+        MOI.ScalarAffineFunction{Float64}(MOI.ScalarAffineTerm{Float64}[], 0.0)
+    model.num_variables_created = 0
+    model.variable_indices = nothing
+    empty!(model.single_variable_mask)
+    empty!(model.lower_bound)
+    empty!(model.upper_bound)
+    empty!(model.var_to_name)
+    model.name_to_var = nothing
+    model.nextconstraintid = 0
+    empty!(model.con_to_name)
+    model.name_to_con = nothing
+    empty!(model.constrmap)
+    MOI.empty!(model.moi_scalaraffinefunction)
```

Comment:
My using `ext` and not `UniversalFallback` ?

Reply:
We would still have the issue with `MOI.empty!`. This is a nicer solution: https://github.com/jump-dev/MathOptInterface.jl/pull/1295

Reply:
Why would we still have the issue? If we start storing part of the model in `ext` then all getters are wrong and we should have `get_without_ext` too

Reply:
I implemented the getters.
https://github.com/jump-dev/AmplNLWriter.jl/blob/e1bdc69b89675744ed93f784cf5d8b79d9b1608b/src/AmplNLWriter.jl#L137-L148

Basically, I want a `Model` that I can store extra stuff in, like the `amplexe` function, and I want to be able to, on `MOI.empty!`, modify some of that extra stuff without deleting it.

Reply:
I guess I could use the current `_NLModel` as `Optimizer`, and only implement `copy_to`.